File: web/src/content.ts
// The content contract shared by the iOS, Android, and web apps.
// Edit ../../content/lessons.json, never the copies the apps render.
import data from "../../content/lessons.json";

export interface Lesson {
  id: string;
  title: string;
  summary: string;
  body: string;
}

export interface QuizQuestion {
  id: string;
  lessonId: string;
  question: string;
  choices: string[];
  answerIndex: number;
  explanation: string;
}

export interface CoinFlipCopy {
  title: string;
  intro: string;
  entangledLabel: string;
  classicalLabel: string;
  measureLabel: string;
  resetLabel: string;
  disclaimer: string;
}

export interface Content {
  app: { name: string; tagline: string; contentVersion: number };
  lessons: Lesson[];
  quiz: QuizQuestion[];
  coinFlip: CoinFlipCopy;
}

export const content: Content = data;

export function lessonById(id: string): Lesson | undefined {
  return content.lessons.find((l) => l.id === id);
}

File: android/app/src/main/kotlin/io/openorcha/entangle/Content.kt
package io.openorcha.entangle

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

// Mirrors content/lessons.json, the contract shared with the iOS and web apps.

@Serializable
data class AppInfo(val name: String, val tagline: String, val contentVersion: Int)

@Serializable
data class Lesson(val id: String, val title: String, val summary: String, val body: String) {
    val paragraphs: List<String> get() = body.split("\n\n")
}

@Serializable
data class QuizQuestion(
    val id: String,
    val lessonId: String,
    val question: String,
    val choices: List<String>,
    val answerIndex: Int,
    val explanation: String,
)

@Serializable
data class CoinFlipCopy(
    val title: String,
    val intro: String,
    val entangledLabel: String,
    val classicalLabel: String,
    val measureLabel: String,
    val resetLabel: String,
    val disclaimer: String,
)

@Serializable
data class Content(
    val app: AppInfo,
    val lessons: List<Lesson>,
    val quiz: List<QuizQuestion>,
    val coinFlip: CoinFlipCopy,
) {
    fun lesson(id: String): Lesson? = lessons.firstOrNull { it.id == id }
}

object ContentStore {
    private val json = Json { ignoreUnknownKeys = true }

    /** lessons.json is packaged as an asset from ../../content by app/build.gradle.kts. */
    fun load(context: Context): Content =
        context.assets.open("lessons.json").bufferedReader().use { json.decodeFromString(Content.serializer(), it.readText()) }
}
